core/species.cpp

```diff
@@ -5486,131 +6101,165 @@ void Species::WriteTreeSequence(std::string &p_recording_tree_path, bool p_simpl
 	// it *does* change the order of the rows; see https://github.com/MesserLab/SLiM/issues/209
 	if (p_simplify)
 	{
-		SimplifyTreeSequence();
+		SimplifyAllTreeSequences();
 	}
 	
-	// Copy the table collection so that modifications we do for writing don't affect the original tables
-	tsk_table_collection_t output_tables;
-	ret = tsk_table_collection_copy(&tables_, &output_tables, 0);
-	if (ret < 0) handle_error("tsk_table_collection_copy", ret);
-	
-	// Sort and deduplicate; we don't need to do this if we simplified above, since simplification does these steps
-	if (!p_simplify)
+	for (Chromosome *chromosome : chromosomes_)
 	{
-		int flags = TSK_NO_CHECK_INTEGRITY;
+		slim_chromosome_index_t chromosome_index = chromosome->Index();
+		TreeSeqInfo &chromosome_tsinfo = treeseq_[chromosome_index];
+		tsk_table_collection_t &chromosome_tables = chromosome_tsinfo.tables_;
+		
+		// Copy in the shared tables (node, individual, population) at this point, so the shared tables then get
+		// copied below; we will be modifying the tables, and don't want our modification to go into the original
+		// shared tables, which we are not allowed to change.
+		if (chromosome_index > 0)
+			CopySharedTablesIn(chromosome_tables);
+		
+		// Copy the table collection so that modifications we do for writing don't affect the original tables.
+		// Note that there's a lot of work below to clean up the individuals table and node table for saving.
+		// Those tables are shared.  We don't want to do this cleanup in the original tables, since that would
+		// modify our recording state I guess; but I think this cleanup will be the same for every chromosome,
+		// so technically we could do this work just once, I think (?), and share the processed tables across
+		// all the chromosomes.  I've chosen not to pursue that idea, because I don't see a path to doing it
+		// without increasing the high-water mark for the memory usage of this code, which is very important
+		// to keep low.  Anyhow, maybe this is unimportant since it is only overhead at save time, and is
+		// probably not a hotspot.
```

Comment:
I agree, probably not a hotspot, but if you do, I don't see anywhere below where there will be an issue:
- sort doesn't touch nodes; and it reorders individuals based only on nodes;
- dedup_sites and compute_mutation_parents don't touch shared tables;
- alive indivs are added to the indiv table, but the same for all

So, it looks to me that if you just move the "add alive individuals" outside the loop over chromosomes, and copy tables over, everything should just work?

Reply:
Huh, OK, I'll look into that, thanks.

Reply:
I can't add alive individuals outside the loop over chromosomes, because then it would be doing it to the original tables (which we don't want to modify) rather than doing it on the copy of the table collection (which is made inside the loop, per chromosome).  So I think this really is trickier to fix that it might appear, and I think/hope the duplicated work is unimportant in the big picture.  Pursue this with me if you disagree.  For now, resolving this.